### PR TITLE
Ordered actions matching

### PR DESCRIPTION
**Pull request by dim0xff**
Matching actions in Catalyst depends on action defining position.
It is OK until you use ActionRoles (eg HTTPMethods). Look:
```
    sub base: Chained('/') PathPart('options') CaptureArgs(0) {}
    sub load: Chained('base') CaptureArgs(1) {}
    sub show: GET OPTIONS Chained('load') Args(0) {}
    sub modify: Chained('load') Args(0) {}
```
POST /options/10 OK (modify)
PUT /options/10 OK (modify)
GET /options/10 OK (show)
OPTIONS /options/10 OK (show)
etc.

BUT! When you rearrange `show` and `modify`, then **ALL** calls will be handled by `modify` action:
```
    sub modify: Chained('load') Args(0) {...}
    sub show: GET OPTIONS Chained('load') Args(0) {...}
```
.
.
So, here is a patch to fix it for `Path` and `Chained`. Matching in `Chained` now depends on position via `Catalyst::Action/compare` (it considers Method, Scheme and Consumes action attributes).

PS: it has side effect: you can apply your own global ActionRole (eg Ordered) with own `compare` method and define matching order as you wish.

PPS: it is fine when you define all your actions in *each* controller. But when you have Base controller, and some controllers differs from Base only on attributes (Chained, PathPart etc, so you manage controller via config) and some controllers have different realization, it is becomes annoying to copy-paste the same code from one controller to another.

### CODE REVIEW

**Comment by jjn1056** [comment]
oh man I am so afraid of this change :)  Yeah this is a problem no doubt.  My big concern is that enabling this is going to potentially break people's stuff, if they had routes setup in a way that worked around the old behavior.  Let me think for bit.  Maybe might need to do this with a feature flag or something.

**Comment by jjn1056** [comment]
something to think about as well, my intention for the next dev cycle is to add matching for Query and Posted params, and also allow one to specify a constraint (like Args(Int) is 1 arg that is an Integer).  I might want to sit on this until that code is in since we'd want to figure out where Post and Query params sit in the order of things.

The issue is the current approach is somewhat unexpected behavior but can be worked around, although its maddening :)  Rather than bake in another new set of assumptions I was thinking maybe we could find a way to 'hint' the dispatcher, to make a certain type of match higher order.  Like you say you can define your own global action role for this as well.

Ultimately I think it makes sense that an action with 'more' match attributes should have higher precedent.  Maybe we can do it this way.  I'm going to cut a development release of Catalyst soonish, maybe even this week.  If you can work with me to get this patch sane with the new type constraint matching I mentioned above we can release it as a dev001 and tell people they gotta test.  and then if they have trouble we can make it a feature flag for back compat.  What do you think?  In general I like where this is going.

**Comment by dim0xff** [comment]
>  I think it makes sense that an action with 'more' match attributes should have higher precedent. 

Maybe. But with **match**ing. How to check - it is matching or not? Example, you can add ActionRoles which just add some stash values and don't perform match against path.

I suggest the next "flow":
1. DispatchTypes must be responsible for matching order and it should be based on `Action/compare`. From docs: compares 2 actions based on the value of the `Args`. With no `Args` having the highest precedence. Fine!
2. Add possibility to add compare rules *(see last two commits)*.
3. If we want back to "old" behaviour, we have to add `postload_dispatch_types` where `compare_rules` is equal to "standard" from Action.pm (default - no sorting, for Args - like in docs: With no `Args` having the highest precedence).

File: lib/Catalyst/DispatchType/Chained.pm [hunk 1]
```diff
@@ -378,7 +378,10 @@ sub register {
 
     $action->attributes->{PathPart} = [ $encoded_part ];
 
-    unshift(@{ $children->{$encoded_part} ||= [] }, $action);
+    # "Reversed" sort: best actions should be at the end
+    $children->{$encoded_part} = [
```
Comment (dim0xff) [inline lib/Catalyst/DispatchType/Chained.pm]: As "semi-update" maybe is reasonable to add method `add_part_action` with default "old" realization:
```
unshift(@{ $children->{$encoded_part} ||= [] }, $action);
```

**Comment by dim0xff** [comment]
I thought a little bit more about it. We have to clarify the problem.

# The problem
Wrong matching path to action (because of addition matching logic in some action roles) in dispatch types *Path* and *Chained*.
Dispatch type matches actions using order which actions was defined in controllers (actually C::DT::Path use sorting by `Args`, but sometimes fails too).
And it can skip addition matching logic:
* for *Path*: when addition match is defined after default
* for *Chained*: when addition match is defined before default

# Matching logic
*Path* and *Chained* actions should respect addition matching logic.
Dispatcher should match path to action with some priority (matcing order).
Each matching order position should depends on `Args` and fact of existing addition matching (in group, like SQL *ORDER BY* by several fields).

## Catalyst::DispatchType::Path
*Path* matching: succeeds on the first action that matches.
Matching order (highest priority on top):
1. Args(*num*). Zero *num* has max priority. Undefined *num* has min priority.
2. Addition matches (like method matching) count: more number of methods is has biggest priority.
### Example
Priority | Args |  Addition matches (count)
---|--------|-------------------------
1 |  Args(0) | Method Scheme (2)
2 |  Args(0) | Method (1)
3 |  Args(0) | Consumes  (1)
4 |  Args(0) | (0)
5 |  Args(0) | (0)
6 |  Args(1) | (0)
7 |  Args(2) | Method Consumes (2)
8 |  Args(2) | (0)
9 |  Args(3) | (0)
10 |  Args(4) | (0)
11 |  Args() | Method (1)
12 |  Args() | (0)

* 2 and 3 - order like its were defined in controller. For 4 and 5 is also. (*)

## Catalyst::DispatchType::Chained
*Chained* matching: succeeds on the last action that matches.
Matching order (highest priority on bottom):
1. Args(*num*). Zero *num* has max priority. Undefined *num* has min priority.
2. Addition matches (like method matching) count: more number of methods is has biggest priority.
### Example
Priority | Args |  Addition matches (count)
---|--------|-------------------------
1 |  Args() | (0)
2 |  Args() | Method (1)
3 |  Args(4) | (0)
4 |  Args(3) | (0)
5 |  Args(2) | (0)
6 |  Args(2) | Method Consumes (2)
7 |  Args(1) | (0)
8 |  Args(0) | (0)
9 |  Args(0) | (0)
10 |  Args(0) | Consumes  (1)
11 |  Args(0) | Method (1)
12 |  Args(0) | Method Scheme (2)

* 10 and 11 - reversed order like its were defined in controller. For 8 and 9 is also. (*)
* So, like **Path**, but reversed.

And yes, you are right with:
> Ultimately I think it makes sense that an action with 'more' match attributes should have higher precedent. 

As conclusion: each ActionRole, which has matching method, should return marker, that it has `match` and priority to sort.
And last commits try to do this (except **(*)**, I think it is reversed). Of course need tests.

**Comment by perl-catalyst-sync** [comment]
hmm, there's a lot going on here...  Ok so lets move this over to the dev
branch (australorp).  Whatever we do it will need to work with the new
action matching stuff we are adding.  Lets ponder it until that stuff lands
since it will likely complicate the picture somewhat

The fact we have so many different dispatch types does help :(  Ideally we
need a good way to expose an interface to allow a programmer to decide.
 -jnap

**Comment by dim0xff** [comment]
> The fact we have so many different dispatch types does help :( Ideally we need a good way to expose an interface to allow a programmer to decide.

Мany? I found 4 (four) in Catalyst distro.
* Path *patched*
* Chained *patched*
* Index. Last modified 5 years ago. Match action with name `index`.
* Default. Last modified 6 years ago. Match Root action with name `default`. Also look [Line#44](https://github.com/perl-catalyst/catalyst-runtime/blob/master/lib/Catalyst/DispatchType/Default.pm#L44) and [Line#54](https://github.com/perl-catalyst/catalyst-runtime/blob/master/lib/Catalyst/DispatchType/Default.pm#L54), does L44 exclude L54?

Seems that Index and Default are being used only when no Path or Chained types are registered.

For all others dispatch types I suggest to use `Catalyst::Action/compare` method (like in Path).
For action roles, which modify `match` method I added some kind of interface to `Catalyst::Action` where you can manage pritority (I also commented it in code)

**Comment by perl-catalyst-sync** [comment]
Dmitry,

I'm hacking on the argument constraint stuff today, lets see if I can merge
those bits with your stuff and then we can see if we like what we have -jnap

**Comment by jjn1056** [comment]
Hey, so I pushed on the first draft of the argument type contraint thing.  Here's a proof of concept commit: https://github.com/perl-catalyst/catalyst-runtime/commit/6d62355b6e08f1234fefbdebaacc9ced07b0bc6c  and I think your work here is going to be pretty important if we want to get this right, so keep with me :)

**Comment by jjn1056** [comment]
ok so we decided for now to stick with the simple 'longest match, last defined rule' and normalize on that.  I'm closing this request but lets keep the idea in mind if we revamp the dispatcher as part of Cat6.

**Pull request closed without merging**